kopf/reactor/watching.py

```diff
@@ -0,0 +1,53 @@
+"""
+Watching & streaming of the watch-events.
+
+Kubernetes client's watching streams are synchronous. To make them asynchronous,
+we put them into a `concurrent.futures.ThreadPoolExecutor`,
+and yield from there asynchronously.
+
+However, async/await coroutines misbehave with `StopIteration` exceptions
+as raise by the `next` method: see `PEP-479`_.
+
+As a workaround, we replace `StopIteration` with our custom `StopStreaming`
+inherited from `RuntimeError` (as suggested by `PEP-479`_),
+and re-implement the generators to make them async.
+
+All of this is a workaround for the standard Kubernetes client's limitations.
+They are not needed if the client library is natively asynchronous.
+
+.. _PEP-479: https://www.python.org/dev/peps/pep-0479/
+"""
+
+import asyncio
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class StopStreaming(RuntimeError):
+    """
+    Raised when the watch-stream generator ends streaming.
+    Replaces `StopIteration`.
+    """
+
+
+def streaming_next(src):
+    """
+    Same as `next`, but replaces the `StopIteration` with `StopStreaming`.
+    """
+    try:
+        return next(src)
+    except StopIteration as e:
+        raise StopStreaming(str(e))
+
+
+async def streaming_aiter(src, loop=None, executor=None):
```

Comment:
"aiter" sounds a bit weird... what about streaming_async_iter?

Reply:
The name is chosen after the existing convention in Python already: `__aiter__`, `__anext__`, `__aenter__`, `__aexit__` — the [magic methods](https://docs.python.org/3/reference/datamodel.html), and the ongoing duscissions all around internet about this method (e.g. [this](https://github.com/python/cpython/pull/8895) or [this](https://bugs.python.org/issue31861) or [PEP-525](https://www.python.org/dev/peps/pep-0525/#aiter-and-anext-builtins) or [google](https://www.google.com/search?q=python+anext)).

So, it seems that `aiter` is the most expected naming convention for this case.

Reply:
aah ok then